### PR TITLE
cpu/stm32f1/periph/adc: ADC implementation for STM32F1

### PR DESCRIPTION
**Pull request by DipSwitch**
The ADC peripheral implementation for the STM32F1xx series.

### CODE REVIEW

**Comment by thomaseichinger** [comment]
@DipSwitch great work. One comment, could you please add braces to [`if`, `for` and `while` blocks](https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions#indentation-and-braces).

**Comment by PeterKietzmann** [comment]
Nice! Sorry to say, but It could take some days until I'll look into this in detail.

**Comment by DipSwitch** [comment]
Fixed braces

**Comment by PeterKietzmann** [comment]
Please add the adc as provided feature like for example [here](https://github.com/RIOT-OS/RIOT/blob/master/boards/stm32f4discovery/Makefile.features#L2)

File: cpu/stm32f1/periph/adc.c [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+ *
+ * @}
+ */
+
+#include "cpu.h"
+#include "periph/adc.h"
+#include "periph_conf.h"
+
+/* guard in case that no ADC device is defined */
+#if ADC_NUMOF
```
Comment (PeterKietzmann) [inline cpu/stm32f1/periph/adc.c]: Where do you define those preprocessor macros? On which board do you test the driver? As far as I see it, the only RIOT supported boards with a stm32f1 CPU are iotlab-m3, spark-core and fox. But you did not add peripheral configurations for any of them

File: cpu/stm32f1/periph/adc.c [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        *port = GPIOB;
+        *pin = channel - 8;
+    }
+    else if (channel < 8) {
+        *port = GPIOA;
+        *pin = channel;
+    }
+    else if (channel > 9) {
+        *port = GPIOC;
+        *pin = channel - 10;
```
Comment (PeterKietzmann) [inline cpu/stm32f1/periph/adc.c]: Wow, that looks weird :-) . Could you point to the regarding section in the reference manual so I don't need to search for it? Would be nice of you

File: cpu/stm32f1/periph/adc.c (hunk 1 above)
Comment (PeterKietzmann) [inline cpu/stm32f1/periph/adc.c]: What do you think, does it make sense to prepare the driver for multiple ADC devices? I'm thinking about something like [this](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32f4/periph/adc.c#L32)

**Comment by PeterKietzmann** [comment]
@DipSwitch, thanks for the implementation. Your code looks really nice! Please answer/address my comments. I will check further details then and hopefully test the driver with an iotlab-m3 board.

File: cpu/stm32f1/periph/adc.c (hunk 1 above)
Comment (DipSwitch) [inline cpu/stm32f1/periph/adc.c]: I test it on a board we want to put on KickStarter before x-mas.

The defines I use in `periph_conf.h` are:

	/**
	 * @name ADC configuration
	 * @{
	 */
	#define ADC_NUMOF           (1U)
	#define ADC_0_EN            1
	#define ADC_MAX_CHANNELS    17

	/* ADC 0 configuration */
	#define ADC_0_DEV           ADC1
	#define ADC_0_CHANNELS      17
	#define ADC_0_CLKEN()       (RCC->APB2ENR |= RCC_APB2ENR_ADC1EN)
	#define ADC_0_CLKDIS()      (RCC->APB2ENR &= ~(RCC_APB2ENR_ADC1EN))
	#define ADC_0_PORT_CLKEN()  (RCC->APB2ENR |= RCC_APB2ENR_IOPAEN | RCC_APB2ENR_IOPBEN | RCC_APB2ENR_IOPCEN)
	/** @} */

The pin defines can be found in in the data sheets of the STM32F1xx MCU's under section 3 (Pinouts and pin description) Table 5. (....-density STM32F103xx pin definitions (continued)) From there out you can search for ADC12_IN##. The idea is that ADC channels 0 -> 7 are mapped on PA0 -> PA7 channels 8 & 9 are mapped on PB0 & PB1 and channels 10 -> 15 on PC0 -> PC5, Channel 16 and 17 are the internal channels for the temperature and the reference. I belief this to be equal for ALL STM's since STM wants to keep there processors drop in replaceable.

Multiple ADC device is possible, but since the read function is blocking the readouts can't interleave I think?

Comment (DipSwitch) [inline cpu/stm32f1/periph/adc.c]: And I just realized that if you readout the ADC from an interrupt, you could corrupt a sample being fetched from a thread. On all STM implementations...

**Comment by PeterKietzmann** [comment]
Now I see 1) you don't need to set the "analog mode" for one pin on the stm32f1 and 2) there is a fix resolution of 12 bit by the ADC. @DipSwitch would you mind adding some documenting lines about the ADC pins to the priph_conf?!

File: cpu/stm32f1/periph/adc.c (hunk 1 above)
Comment (PeterKietzmann) [inline cpu/stm32f1/periph/adc.c]: @DipSwitch I currently don't have enough time to spend in this. If you're fine we can go on with this PR for now and you open an issue for the general question

File: cpu/stm32f1/periph/adc.c [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+#include "cpu.h"
+#include "periph/adc.h"
+#include "periph_conf.h"
+
+/* guard in case that no ADC device is defined */
+#if ADC_NUMOF
+
+static bool _adc_get_pin(int channel, GPIO_TypeDef * *port, int *pin)
+{
+    *port = NULL;
```
Comment (PeterKietzmann) [inline cpu/stm32f1/periph/adc.c]: I get ` makes integer from pointer without a cast [-Werror]` here (with gcc version 4.8.4) . Maybe just do a void cast?

**Comment by PeterKietzmann** [comment]
I am thinking about the need of a `ADC_0_CH0_PIN` pin from the periph_conf file. For example, the iotlab-m3 board has just few pins. The only ADC relevant pin there is PA3 which correlates with ADC channel 3. But I think from the RIOT perspective one might want to call it channel 0 or 1. @haukepetersen spent more time in this pin configuration things. Also it might be time for the "ADC interface optimization" like already done for GPIO and started for SPI and PWM

**Comment by DipSwitch** [comment]
Fixed the build error. Squash?

**Comment by PeterKietzmann** [comment]
@DipSwitch could you please add an appropriate periph_conf entry for the iotlab-m3 board, like you did for the PWM driver? The relevant pin would be PA3. Ok to squash

**Comment by DipSwitch** [comment]
Added, but note that the driver currently "steals" the pin. The user application must make sure to only call `adc_read(3)`. We could add something like:

    /* define ADC channels */
    #define ADC_CHANNEL_0       3

**Comment by PeterKietzmann** [comment]
Sry, I don't understand what you're saying. But maybe my next comment (inline) will clear it up anyway

File: boards/iotlab-m3/include/periph_conf.h [hunk 4]
```diff
@@ -43,6 +43,25 @@ extern "C" {
...
+ * @name ADC configuration
+ * @{
+ */
+#define ADC_NUMOF           (1U)
+#define ADC_0_EN            1
+#define ADC_MAX_CHANNELS    17
+
+/* ADC 0 configuration */
+#define ADC_0_DEV           ADC1
+#define ADC_0_CHANNELS      17
```
Comment (PeterKietzmann) [inline boards/iotlab-m3/include/periph_conf.h]: I think as long as we just have the ability to use one adc pin on the iotlab-m3 board we should define that macro with `1`

File: boards/iotlab-m3/include/periph_conf.h [hunk 5]
```diff
@@ -43,6 +43,25 @@ extern "C" {
...
+#define ADC_0_CLKEN()       (RCC->APB2ENR |= RCC_APB2ENR_ADC1EN)
+#define ADC_0_CLKDIS()      (RCC->APB2ENR &= ~(RCC_APB2ENR_ADC1EN))
+#define ADC_0_PORT_CLKEN()  (RCC->APB2ENR |= RCC_APB2ENR_IOPAEN | RCC_APB2ENR_IOPBEN | RCC_APB2ENR_IOPCEN)
+
+/* define ADC channels */
+#define ADC_CHANNEL_0       3
```
Comment (PeterKietzmann) [inline boards/iotlab-m3/include/periph_conf.h]: Why not defining it like already done for other peripheral configurations, e.g. like [this one](https://github.com/RIOT-OS/RIOT/blob/master/boards/stm32f4discovery/include/periph_conf.h#L131)

Comment (DipSwitch) [inline boards/iotlab-m3/include/periph_conf.h]: Then I have a better solution I think. mom

**Comment by DipSwitch** [comment]
Maybe something like this is better? Maybe it can also be ported to other boards?

**Comment by PeterKietzmann** [comment]
@DipSwitch, your proposal looks reasonable on the first sight, but this is @haukepetersen's expertise.

**Comment by PeterKietzmann** [comment]
hups, clicked too early :-)

**Comment by PeterKietzmann** [comment]
So, as this is Haukes expertise, I don't want to cross his plans and would wait for his feedback. How urgent is it for this driver to get into master? In case it *is* urgent I tend to go for my solution (cause we're currently all busy and the feature-freeze stands before). Otherwise it would be a nice improvement which might take some days longer.

**Comment by DipSwitch** [comment]
It's not that of urgent if you ask me :) The board which is gonna use it isn't gonna be added yet.

**Comment by PeterKietzmann** [comment]
Fine, then let's wait to hear Hauke's opinion and go on then

**Comment by haukepetersen** [comment]
I think this approach goes in the right direction, but is not 'quite there' yet. I don't really like the method the ADC pins are defined. This should be done somehow in the `periph_conf.h`. Those `xx_params.h` files are rather for configuring device drivers, network devices etc, but not on-cpu pin mappings.

I have something in mind that might work and fits in the exsiting paradigm: I will create a quick PR against you branch.

**Comment by haukepetersen** [comment]
Oh, and I just noticed one missunderstanding: the `channel` variable in the ADC interface is meant as logical channel, it is not connected to any channels that the CPU uses internally - so if you configure 1 channel for `ADC_0`, it is automatically channel `0`. So it is up to the periph driver (and the `periph_conf.h`, to map the STM32F1s internal channels to logical channels starting from 0...

**Comment by haukepetersen** [comment]
Ok, took a moment, I decided against a PR against this branch, but I proposed an API change and made a sample implementation for the `STM32F4` instead: #4430

The implementation for the F1 should look then very much similar to the `F4's` implementation - if you accept my changes :-)

**Comment by DipSwitch** [comment]
Needs label Wait for Other PR

**Comment by OlegHahm** [comment]
@DipSwitch, what's the state? Still waiting for #4430?

**Comment by DipSwitch** [comment]
Yes, but not for long I guess ;) will merge squash and rebase once it's merged.

**Comment by DipSwitch** [comment]
Should be ready for merge.

@haukepetersen @PeterKietzmann agree?

File: cpu/stm32f1/include/stm32f103xe.h [hunk 6]
```diff
@@ -11,7 +11,7 @@
   *          This file contains:
   *           - Data structures and the address mapping for all peripherals
   *           - Peripheral's registers declarations and bits definition
-  *           - Macros to access peripheral�s registers hardware
+  *           - Macros to access peripheral’s registers hardware
```
Comment (haukepetersen) [inline cpu/stm32f1/include/stm32f103xe.h]: unrelated?

**Comment by haukepetersen** [comment]
nice, looks good to me, though I have not tested it -> untested ACK, will give it a try tomorrow.

File: cpu/stm32f1/include/stm32f103xe.h (hunk 6 above)
Comment (DipSwitch) [inline cpu/stm32f1/include/stm32f103xe.h]: addressed and squashed.

**Comment by haukepetersen** [comment]
successfully tested on the `iotlab-m3`, ACK and go.
